### DISCUSSION TITLE
[SDK] Python SDK connection to web socket endpoint

### DISCUSSION THREAD

**Discussion by enochSatis** [zkSync-Community-Hub/zksync-developers#189] [SDKs] [answered]
### Environment

Mainnet, zksync2 v.0.4 in Python

### Acknowledgement

- [X] I understand that issues are expected with the Go SDK as they are being updated.

### Issue Description

I would like to use the `get_filter_logs` method in web3. When I am using `ZkSyncBuilder` in `zksync2`, seems that the connection to the mainnet websocket endpoint is refused. When I am using the http endpoint, the SDK returns `Filter not found` error. 

### Expected Behavior

1. Can filter be constructed with an http endpoint in zkSync? I get `{'code': -32602, 'message': 'Filter not found'}` when I use the http endpoint. 
2. Can `zksync2` Python SDK connect to websocket endpoint? I get `No connection adapters were found for 'wss://mainnet.era.zksync.io/ws'` when I tried to connect to the websocket endpoint. 

Seems the examples on gitHub are using http endpoints. 

### Code Example

```
from zksync2.module.module_builder import ZkSyncBuilder # zksync's provider
from zksync2.core.types import EthBlockParams # zksync requirement to get transaction count
from zksync2.manage_contracts.contract_encoder_base import ContractEncoder # zksync requirement to create contract
from pathlib import Path # zksync ABI file needs Path object instead of string input


node = ZkSyncBuilder.build("wss://mainnet.era.zksync.io/ws")
#node = ZkSyncBuilder.build("https://mainnet.era.zksync.io/")
contract_addr = "0x42dCF68ea66be9890BC233A5A7989Cc1b2845E3a"


event_filter = node.zksync.filter({"address": contract_addr})
log_res = node.zksync.get_filter_logs(event_filter.filter_id)
print(log_res)
```

**Comment by enochSatis** [comment, marked as answer]
Found a way to solve this issue. 

I got the logs by calling tPOST API of the RPC node directly using method `eth_getLogs`. 

```
"id": 2,
"jsonrpc": "2.0",
"method": "eth_getLogs",
"params": [{
    "address": address,
    "topics": [topicA, topicB]
    }]
```

**Comment by daisbryce** [comment]
It seems that the zksync2 Python SDK currently does not support connecting to the websocket endpoint or constructing filters with an HTTP endpoint. The Filter not found error you're encountering suggests that the SDK is not able to find the filter you're trying to use.

You may need to wait for updates or enhancements to the SDK that provide support for these features. In the meantime, you can consider reaching out to the zkSync team for further assistance or exploring alternative solutions for your requirements.

**Comment by etesvipatdeb** [comment]
# Ensure you are using the correct endpoint format for the ZkSyncBuilder in zksync2 Python SDK.
# For websocket connection, use 'ws' instead of 'wss'.
# For HTTP connection, use 'http' instead of 'https'.
# Here's the corrected code:

from zksync2.module.module_builder import ZkSyncBuilder
from zksync2.core.types import EthBlockParams

 Connect using websocket endpoint
node_ws = ZkSyncBuilder.build("ws://mainnet.era.zksync.io/ws")

 Connect using HTTP endpoint
node_http = ZkSyncBuilder.build("http://mainnet.era.zksync.io/")

contract_addr = "0x42dCF68ea66be9890BC233A5A7989Cc1b2845E3a"

 Create filter
event_filter = node_ws.zksync.filter({"address": contract_addr})

 Get filter logs
log_res = node_ws.zksync.get_filter_logs(event_filter.filter_id)
print(log_res)
Ensure you use the correct endpoint format and protocol (ws or http) for the ZkSyncBuilder connection. Also, make sure the contract address is correct and that the event filter is properly constructed.

**Comment by poinsondrew** [comment]
Use HTTP endpoint for now to avoid "Filter not found" error
node = ZkSyncBuilder.build("https://mainnet.era.zksync.io/")
 Alternatively, use the correct syntax for connecting to the websocket endpoint
 node = ZkSyncBuilder.build("wss://mainnet.era.zksync.io")

**Comment by laducerbecky** [comment]
It seems that the zksync2 Python SDK is currently unable to connect to the mainnet websocket endpoint, and using the http endpoint returns a "Filter not found" error when attempting to construct a filter with the get_filter_logs method. The examples on GitHub also appear to be using http endpoints.

**Comment by hilliardloyd** [comment]
Currently, the zksync2 Python SDK does not support WebSocket endpoints. Therefore, you will need to use an HTTP endpoint to construct a filter. However, the "Filter not found" error suggests that the filter you are trying to access does not exist. Make sure that the contract address you are using is correct and that the event you are filtering for has occurred.
